halibot/halauth.py

```diff
@@ -2,15 +2,15 @@
 import logging
 from halibot.message import Message
 
-def hasPermission(perm, reply=False):
+def hasPermission(perm, reply=False, argnum=-1, key="msg"):
 	def real_dec(func):
 		def wrapper(self, *args, **kwargs):
-			msg = None
-			for i in list(args) + list(kwargs.values()):
-				if i.__class__ == Message:
-					msg = i
-					break
+			if argnum >= 0 and argnum < len(args):
+				msg = args[argnum]
 			else:
+				msg = kwargs.get(key)
+
+			if not msg:
 				self.log.error("Probable module bug! -- hasPermission decorator called on a function that doesn't have a Message argument!")
```

Comment:
Not necessarily true: it could have successfully retrieved a message from the kwargs that was set to a false value. Either change the log or make the check `if not key in kwargs`.

Reply:
What about `if not (hasattr(msg, "origin") and `hasattr(msg, "identity"))` ? That's all we are really looking to extract from the message object anyway, less chance for false positives.

Reply:
What about just catching AttributeErrors for those when we try to use them?

Reply:
I guess that surfaces less to the module. Erg. Yeah, we should probably just check for those attrs.

Reply:
Actually, I just added a `MalformedMsgException` in #131 , maybe we should utilize that here too? If attr check fails raise that complaining about the bad attr.

Reply:
Seems reasonable. Should we throw that where we use those attrs then? It would keep that which enforces constraints and the reason for them together.

Reply:
We probably should, but not as a `.__getattr__()` action. There's always a possibility that we are handed a non-Message object, so we should probably check those attrs where it is critical to do so. That is to say, `.raw_send()` should, and auth checks should.

Reply:
Something I just thought of here, should we handle the case of `key not in kwargs.keys()` separately from `kwargs[key] = None/Badobject`?